File: attestations/LLM_Attestation.js
// NVIDIA GPU Attestation Verification Script (Fixed for JWT responses)
// This script properly handles JWT-based attestation responses

const API_KEY = 'sk-cb209387ad164ffb8aedd7a0d3ab57dc';
const BASE_URL = 'https://865d4c24102519d7d0a91f5918902ff1ca4b9670-3000.dstack-prod5.phala.network';
const MODEL = 'openai/gpt-oss-120b';
const NVIDIA_ATTESTATION_URL = 'https://nras.attestation.nvidia.com/v3/attest/gpu';

// Function to decode JWT without verification (just to read the payload)
function decodeJWT(token) {
  try {
    const parts = token.split('.');
    if (parts.length !== 3) return null;
    
    // Decode the payload (second part)
    const payload = parts[1];
    const decoded = Buffer.from(payload, 'base64').toString('utf-8');
    return JSON.parse(decoded);
  } catch (error) {
    console.error('Error decoding JWT:', error.message);
    return null;
  }
}

async function getAttestationReport() {
  console.log('📥 Fetching attestation report...\n');
  
  const response = await fetch(
    `${BASE_URL}/v1/attestation/report?model=${encodeURIComponent(MODEL)}`,
    {
      headers: {
        'Authorization': `Bearer ${API_KEY}`
      }
    }
  );

  if (!response.ok) {
    throw new Error(`Failed to fetch attestation: ${response.status} ${response.statusText}`);
  }

  return await response.json();
}

async function verifyNvidiaGPU(nvidiaPayload) {
  console.log('🔐 Verifying NVIDIA GPU attestation...\n');
  
  // Parse the nvidia_payload string to JSON
  const evidence = JSON.parse(nvidiaPayload);
  
  const response = await fetch(NVIDIA_ATTESTATION_URL, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(evidence)
  });

  const status = response.status;
  const responseBody = await response.text();
  
  // Try to parse as JSON
  let parsedBody;
  try {
    parsedBody = JSON.parse(responseBody);
  } catch (e) {
    return {
      verified: false,
      status: status,
      error: 'Failed to parse response',
      response: responseBody
    };
  }

  // NVIDIA returns an array: [["JWT", "token..."], {"GPU-0": "token..."}]
  // We need to decode the first JWT to get x-nvidia-overall-att-result
  
  let overallResult = null;
  let jwtPayload = null;
  
  // Check if response is an array with JWT
  if (Array.isArray(parsedBody) && parsedBody.length > 0) {
    const firstElement = parsedBody[0];
    
    if (Array.isArray(firstElement) && firstElement[0] === 'JWT' && firstElement[1]) {
      const jwtToken = firstElement[1];
      jwtPayload = decodeJWT(jwtToken);
      
      if (jwtPayload && jwtPayload['x-nvidia-overall-att-result'] !== undefined) {
        overallResult = jwtPayload['x-nvidia-overall-att-result'];
      }
    }
  }
  
  // Also check the old header method (fallback)
  const headerResult = response.headers.get('x-nvidia-overall-att-result');
  if (headerResult) {
    overallResult = headerResult === 'True' || headerResult === 'true';
  }
  
  const verified = overallResult === true || overallResult === 'True' || overallResult === 'true';
  
  return {
    verified: verified,
    status: status,
    overallResult: overallResult,
    jwtPayload: jwtPayload,
    response: responseBody
  };
}

async function main() {
  try {
    console.log('═══════════════════════════════════════════════════');
    console.log('  NVIDIA GPU TEE Attestation Verification');
    console.log('═══════════════════════════════════════════════════\n');

    // Step 1: Get attestation report
    const attestation = await getAttestationReport();
    
    console.log('✅ Attestation report received');
    console.log(`   Signing Address: ${attestation.signing_address}`);
    console.log(`   Signing Algorithm: ${attestation.signing_algo}`);
    console.log(`   App ID: ${attestation.info.app_id}`);
    console.log(`   Instance ID: ${attestation.info.instance_id}\n`);

    // Step 2: Verify NVIDIA GPU
    const nvidiaResult = await verifyNvidiaGPU(attestation.nvidia_payload);
    
    console.log('═══════════════════════════════════════════════════');
    console.log('  NVIDIA GPU Verification Result');
    console.log('═══════════════════════════════════════════════════\n');

    console.log(`NVIDIA GPU Attestation Details:${nvidiaResult}`);
    console.log(`NVIDIA GPU Attestation Details:${nvidiaResult.response}`);
    console.log(`NVIDIA GPU Attestation Details:${nvidiaResult.verified}`);
    
    if (nvidiaResult.verified) {
      console.log('✅ NVIDIA GPU VERIFICATION PASSED');
      console.log('   Status: Genuine NVIDIA Hardware in TEE Mode');
      console.log(`   HTTP Status: ${nvidiaResult.status}`);
      console.log(`   Overall Result: ${nvidiaResult.overallResult}`);
      
      if (nvidiaResult.jwtPayload) {
        console.log('\n📋 JWT Payload Details:');
        console.log(`   Issuer: ${nvidiaResult.jwtPayload.iss || 'N/A'}`);
        console.log(`   Subject: ${nvidiaResult.jwtPayload.sub || 'N/A'}`);
        console.log(`   Version: ${nvidiaResult.jwtPayload['x-nvidia-ver'] || 'N/A'}`);
        console.log(`   Issued At: ${new Date(nvidiaResult.jwtPayload.iat * 1000).toISOString()}`);
        console.log(`   Expires At: ${new Date(nvidiaResult.jwtPayload.exp * 1000).toISOString()}`);
      }
    } else {
      console.log('❌ NVIDIA GPU VERIFICATION FAILED');
      console.log(`   HTTP Status: ${nvidiaResult.status}`);
      console.log(`   Overall Result: ${nvidiaResult.overallResult}`);
      
      if (nvidiaResult.error) {
        console.log(`   Error: ${nvidiaResult.error}`);
      }
      
      if (nvidiaResult.jwtPayload) {
        console.log('\n📋 JWT Payload (for debugging):');
        console.log(JSON.stringify(nvidiaResult.jwtPayload, null, 2));
      }
    }
    
    console.log('\n═══════════════════════════════════════════════════\n');
    
    // Return verification result
    return {
      success: nvidiaResult.verified,
      attestation: {
        signingAddress: attestation.signing_address,
        appId: attestation.info.app_id,
        instanceId: attestation.info.instance_id
      },
      nvidia: {
        verified: nvidiaResult.verified,
        status: nvidiaResult.status,
        overallResult: nvidiaResult.overallResult
      }
    };

  } catch (error) {
    console.error('❌ Error during verification:', error.message);
    console.error(error);
    throw error;
  }
}

// Run the verification
main()
  .then(result => {
    console.log('Verification completed successfully');
    console.log(`Final Result: ${result.success ? '✅ PASSED' : '❌ FAILED'}`);
    process.exit(result.success ? 0 : 1);
  })
  .catch(error => {
    console.error('Verification failed');
    process.exit(1);
  });

File: attestations/LLM_Attestation.py
#!/usr/bin/env python3
"""Complete LLM Verification: Chat Completion + TEE Attestation + Signature"""

import requests
import json
import base64
from eth_account.messages import encode_defunct
from eth_account import Account

API_KEY = 'sk-cb209387ad164ffb8aedd7a0d3ab57dc'
BASE_URL = 'https://865d4c24102519d7d0a91f5918902ff1ca4b9670-3000.dstack-prod5.phala.network'
MODEL = 'openai/gpt-oss-120b'
NVIDIA_URL = 'https://nras.attestation.nvidia.com/v3/attest/gpu'


def decode_jwt(token):
    """Decode JWT payload"""
    try:
        parts = token.split('.')
        if len(parts) != 3:
            return None
        payload = parts[1]
        padding = 4 - len(payload) % 4
        if padding != 4:
            payload += '=' * padding
        decoded = base64.urlsafe_b64decode(payload)
        return json.loads(decoded)
    except Exception as e:
        print(f'JWT decode error: {e}')
        return None


def chat_completion(message):
    """Step 1: Chat completion"""
    response = requests.post(
        f'{BASE_URL}/v1/chat/completions',
        headers={'Authorization': f'Bearer {API_KEY}', 'Content-Type': 'application/json'},
        json={'model': MODEL, 'messages': [{'role': 'user', 'content': message}]}
    )
    return response.json()


def get_tee_report():
    """Step 2: Get TEE attestation report"""
    response = requests.get(
        f'{BASE_URL}/v1/attestation/report',
        headers={'Authorization': f'Bearer {API_KEY}'},
        params={'model': MODEL}
    )
    return response.json()


def get_signature(completion_id):
    """Step 4: Get signature for completion"""
    response = requests.get(
        f'{BASE_URL}/v1/signature/{completion_id}',
        headers={'Authorization': f'Bearer {API_KEY}'},
        params={'model': MODEL, 'signing_algo': 'ecdsa'}
    )
    return response.json()


def verify_nvidia_attestation(nvidia_payload):
    """Step 5: Verify NVIDIA TEE attestation"""
    evidence = json.loads(nvidia_payload)
    response = requests.post(
        NVIDIA_URL,
        headers={'Content-Type': 'application/json'},
        json=evidence
    )
    
    try:
        parsed_body = response.json()
        if isinstance(parsed_body, list) and len(parsed_body) > 0:
            first_element = parsed_body[0]
            if isinstance(first_element, list) and first_element[0] == 'JWT':
                jwt_payload = decode_jwt(first_element[1])
                if jwt_payload and 'x-nvidia-overall-att-result' in jwt_payload:
                    return jwt_payload['x-nvidia-overall-att-result'] is True
    except Exception as e:
        print(f'NVIDIA verification error: {e}')
    
    return False


def verify_signature(text_hash, signature):
    """Step 6: Verify signature and recover address"""
    message = encode_defunct(text=text_hash)
    recovered_address = Account.recover_message(message, signature=signature)
    return recovered_address


def main():
    print('=' * 60)
    print('COMPLETE LLM VERIFICATION CHAIN')
    print('=' * 60)
    
    # Step 1: Chat completion
    print('\n[1/6] Chat completion...')
    completion = chat_completion('What is 2+2?')
    completion_id = completion['id']
    response_text = completion['choices'][0]['message']['content']
    print(f'✓ Completion ID: {completion_id}')
    print(f'✓ Response: {response_text[:80]}...')
    
    # Step 2: Get TEE report
    print('\n[2/6] Fetching TEE attestation report...')
    attestation = get_tee_report()
    tee_signing_address = attestation['signing_address']
    nvidia_payload = attestation['nvidia_payload']
    print(f'✓ TEE Signing Address: {tee_signing_address}')
    
    # Step 3: Store TEE signing address
    print('\n[3/6] Storing TEE signing address...')
    stored_tee_address = tee_signing_address
    print(f'✓ Stored: {stored_tee_address}')
    
    # Step 4: Get signing response
    print('\n[4/6] Fetching signature for completion...')
    sig_data = get_signature(completion_id)
    text_hash = sig_data['text']
    signature = sig_data['signature']
    sig_response_address = sig_data['signing_address']
    print(f'✓ Text Hash: {text_hash}')
    print(f'✓ Signature: {signature}')
    print(f'✓ Signature Response Address: {sig_response_address}')
    
    # Step 5: Verify TEE report (NVIDIA attestation)
    print('\n[5/6] Verifying NVIDIA GPU attestation...')
    nvidia_verified = verify_nvidia_attestation(nvidia_payload)
    if nvidia_verified:
        print('✓ NVIDIA Attestation: PASSED (Genuine NVIDIA hardware in TEE)')
    else:
        print('✗ NVIDIA Attestation: FAILED')
        return False
    
    # Step 6: Verify signature
    print('\n[6/6] Verifying signature...')
    recovered_address = verify_signature(text_hash, signature)
    print(f'✓ Recovered Address: {recovered_address}')
    
    # Compare all 3 addresses
    print('\n' + '=' * 60)
    print('ADDRESS VERIFICATION')
    print('=' * 60)
    print(f'TEE Signing Address (Step 3):      {stored_tee_address}')
    print(f'Signature Response Address (Step 4): {sig_response_address}')
    print(f'Recovered Address (Step 6):          {recovered_address}')
    
    match_1 = stored_tee_address.lower() == sig_response_address.lower()
    match_2 = stored_tee_address.lower() == recovered_address.lower()
    match_3 = sig_response_address.lower() == recovered_address.lower()
    
    print(f'\nTEE ↔ Signature Response: {"✓ MATCH" if match_1 else "✗ MISMATCH"}')
    print(f'TEE ↔ Recovered:          {"✓ MATCH" if match_2 else "✗ MISMATCH"}')
    print(f'Signature ↔ Recovered:    {"✓ MATCH" if match_3 else "✗ MISMATCH"}')
    
    all_match = match_1 and match_2 and match_3
    
    print('\n' + '=' * 60)
    print('FINAL VERIFICATION RESULT')
    print('=' * 60)
    if nvidia_verified and all_match:
        print('✅ COMPLETE VERIFICATION PASSED')
        print('   • Hardware: Genuine NVIDIA GPU in TEE')
        print('   • Signature: Valid and matches TEE instance')
        print('   • Chain of Trust: VERIFIED')
        return True
    else:
        print('❌ VERIFICATION FAILED')
        if not nvidia_verified:
            print('   • NVIDIA attestation failed')
        if not all_match:
            print('   • Address mismatch detected')
        return False


if __name__ == '__main__':
    success = main()
    exit(0 if success else 1)